# TP 1 Enunciado

Tareas
El equipo debe realizar las siguientes actividades.
1. Identificar el tipo/clase de problema.
2. Identificar las variables y sus características. Describirlas.
3. Realizar la tarea de preprocesamiento en caso de que fuese necesario. Limpieza de variables,
normalización, categorización, remuestreo, binarización, categorización, etc.
4. Aplicar al menos cuatro (4) modelos de Machine Learning distintos para el problema
identificado.
5. Formalizar una técnica de selección de los datos.
6. Evaluar cada modelo según los resultados obtenidos.
7. Comentar las ventajas y limitación de los modelos elegidos.
8. Presentar los resultados obtenidos.

## Import de Libs

In [28]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

import pandas as pd

## Leer datos de .csv

In [29]:
df = pd.read_csv("Churners.csv")

## Analisis del dataset

In [30]:
print("Tamaño del dataset:",len(df));

Tamaño del dataset: 9000


In [31]:
print("Datos: ", df.keys())

Datos:  Index(['id', 'Nombre', 'Email', 'Pais', 'Genero', 'Edad', 'Antigüedad',
       'FacturacionTotal', 'PuntuacionCrediticia', 'CantidadProductos',
       'PoseeTarjetaCredito', 'MiembroActivo', 'SalarioEstimado',
       'VelocidadServicio', 'Abandono'],
      dtype='object')


In [32]:
df.head()

id                     Nombre                             Email  \
0  15638513       Estrella de Alvarado          roxanafuster@rodenas.org   
1  15805303   Dora Nieves Lerma Zurita              maximino90@gmail.com   
2  15685844  Adelina Contreras Cervera          somozaeufemia@revilla.es   
3  15658693            Ale del Portero                gisela65@amores.es   
4  15676091      Salvador Ángel Gimeno  maximino64@navarrete-gonzalo.com   

        Pais     Genero  Edad  Antigüedad  FacturacionTotal  \
0  Argentina   Femenino    40           7         142856.95   
1     Brasil  Masculino    44           1         141136.62   
2     Brasil   Femenino    35           8         141665.63   
3  Argentina   Femenino    60           2              0.00   
4  Argentina  Masculino    42           7              0.00   

   PuntuacionCrediticia  CantidadProductos PoseeTarjetaCredito MiembroActivo  \
0                   723                  2                  No            No   
1                   661                  1                  Si            No   
2                   518                  1                  No            Si   
3                   827                  2                  No            Si   
4                   543                  1                  Si            Si   

   SalarioEstimado VelocidadServicio Abandono  
0         38019.74             500Mb       Si  
1        189742.78             500Mb       Si  
2        192776.64               1GB       No  
3         60615.83               1GB       No  
4         56650.47             500Mb       Si

In [33]:
df.describe()

id         Edad   Antigüedad  FacturacionTotal  \
count  9.000000e+03  9000.000000  9000.000000       9000.000000   
mean   1.569074e+07    38.944333     5.021889      76401.613553   
std    7.175292e+04    10.489313     2.895496      62372.158110   
min    1.556570e+07    18.000000     0.000000          0.000000   
25%    1.562829e+07    32.000000     3.000000          0.000000   
50%    1.569049e+07    37.000000     5.000000      97089.635000   
75%    1.575281e+07    44.000000     8.000000     127651.035000   
max    1.581569e+07    92.000000    10.000000     250898.090000   

       PuntuacionCrediticia  CantidadProductos  SalarioEstimado  
count           9000.000000        9000.000000      9000.000000  
mean             651.088111           1.530667    100171.664682  
std               96.659877           0.581366     57574.932752  
min              350.000000           1.000000        11.580000  
25%              584.000000           1.000000     51078.675000  
50%              652.000000           1.000000    100185.240000  
75%              718.000000           2.000000    149577.610000  
max              850.000000           4.000000    199992.480000

In [34]:
df["Genero"].value_counts()

Masculino    4923
Femenino     4077
Name: Genero, dtype: int64

In [35]:
df["MiembroActivo"].value_counts()

Si    4631
No    4369
Name: MiembroActivo, dtype: int64

In [36]:
df["Abandono"].value_counts()

No    5753
Si    3247
Name: Abandono, dtype: int64

In [37]:
df["Pais"].value_counts()

Argentina    4523
Brasil       2267
Chile        2210
Name: Pais, dtype: int64

In [38]:
df["VelocidadServicio"].value_counts()

20GB     1842
1GB      1807
500Mb    1793
2GB      1788
100GB    1770
Name: VelocidadServicio, dtype: int64

# Resumen

* Rango edad: 18 - 92
* Antiguedad: 0 - 10
* Productos: 1 - 4
* Salario: 11.580000 - 199992.480000

# Limpieza de datos

In [39]:
# Valores categoricos a codigo 1 / 0 / etc
df["Abandono"] = pd.Categorical(df.Abandono).codes
df["Genero"] = pd.Categorical(df.Genero).codes
df["Pais"] = pd.Categorical(df.Pais).codes
df["MiembroActivo"] = pd.Categorical(df.MiembroActivo).codes
df["VelocidadServicio"] = pd.Categorical(df.VelocidadServicio).codes
df["PoseeTarjetaCredito"] = pd.Categorical(df.PoseeTarjetaCredito).codes
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 15 columns):
id                      9000 non-null int64
Nombre                  9000 non-null object
Email                   9000 non-null object
Pais                    9000 non-null int8
Genero                  9000 non-null int8
Edad                    9000 non-null int64
Antigüedad              9000 non-null int64
FacturacionTotal        9000 non-null float64
PuntuacionCrediticia    9000 non-null int64
CantidadProductos       9000 non-null int64
PoseeTarjetaCredito     9000 non-null int8
MiembroActivo           9000 non-null int8
SalarioEstimado         9000 non-null float64
VelocidadServicio       9000 non-null int8
Abandono                9000 non-null int8
dtypes: float64(2), int64(5), int8(6), object(2)
memory usage: 685.7+ KB


# Modelo de ML

In [40]:
data = df.loc[:, df.columns != 'Abandono']
data = data.loc[:, data.columns != 'Nombre']
data = data.loc[:, data.columns != 'Email']
target = df["Abandono"]

In [41]:
from sklearn.model_selection import train_test_split

In [42]:
X_train, X_test, Y_train, Y_test = train_test_split(data, target, random_state=0)

In [ ]:
pd.plotting.scatter_matrix(X_train, c=Y_train, figsize=(15,15), marker='o', hist_kwds={'bins':20}, s=60, alpha=8)

array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7fa0f279bbd0>,
      dtype=object)

## Pruebas con cada Modelo

In [ ]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score,roc_curve,confusion_matrix
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error

# Regesion Lineal

In [ ]:
from sklearn.linear_model import LinearRegression

lr_model = LinearRegression()
lr_model.fit(X_train,Y_train)
Y_pred = lr_model.predict(X_test)

print('Test Score:', lr_model.score(X_test,Y_test))
print('Train Score:', lr_model.score(X_train,Y_train))
y_pred = lr_model.predict(X_test)
r2score = r2_score(Y_test,y_pred)
print("R2 Score:", r2score)

# Random Forest

In [ ]:
from sklearn.ensemble import RandomForestRegressor

rf_model = RandomForestRegressor()
rf_model.fit(X_train, Y_train)

print('Test Score:', rf_model.score(X_test,Y_test))
print('Train Score:', rf_model.score(X_train,Y_train))
y_pred = rf_model.predict(X_test)
r2score = r2_score(Y_test,y_pred)
print("R2 Score:", r2score)

# KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

kn_model = KNeighborsClassifier(n_neighbors=2)
kn_model.fit(X_train, Y_train)

print('Test Score:', kn_model.score(X_test,Y_test))
print('Train Score:', kn_model.score(X_train,Y_train))
y_pred = kn_model.predict(X_test)
r2score = r2_score(Y_test,y_pred)
print("R2 Score:", r2score)

# Support Vector Machine

In [ ]:
from sklearn.svm import SVC

svm_model = SVC()
svm_model.fit(X_train, Y_train)

print('Test Score:', svm_model.score(X_test,Y_test))
print('Train Score:', svm_model.score(X_train,Y_train))
y_pred = svm_model.predict(X_test)
r2score = r2_score(Y_test,y_pred)
print("R2 Score:", r2score)